Doc/library/stdtypes.rst

```diff
@@ -2392,6 +2392,196 @@ that ``'\0'`` is the end of the string.
+
+   >>> name = "Fred"
+   >>> f"He said his name is {name!r}."
+   "He said his name is 'Fred'."
+
+The above is an example and will become clear as you read below.
+
+Everything outside the curly braces is treated as literally with the exception
+of double braces which are treated as one brace::
+
+   >>> f"This is a string"
+   'This is a string.'
+   >>> f'{{'
+   '{'
+   >>> F"{{"
+   '}'
+   >>> F'{{ }}'
+   '{ }'
+
+A single opening curly bracket, ``'{'``, marks a replacement field that starts with
+a python expression.::
+
+   >>> f"4 * 3 is {4 * 3}"
+   '4 * 3 is 12'
+
+The above is quite contrived and a more useful use in code could be::
+
+   >>> a = 4
+   >>> b = 3
+   >>> f"The product of {a} and {b} is {a * b}"
+   'The product of 4 and 3 is 12'
+
+Or more simply::
+
+   >>> a = 4
+   >>> b = 3
+   >>> f"The product is {a * b}"
+   'The product is 12'
+
+Methods can also be used::
+
+   >>> string = "hello, world!"
+   >>> f"{string.title()}"
+   'Hello, World!'
+   >>> def make_title(input):
+   ...     return input.title()
+   ...
+   >>> f"{make_title(string)}"
+   'Hello, World!'
+
+Any variable that can be presented as a string can be used with f-strings::
+
+   >>> import datetime
+   >>> now = datetime.datetime.now()
+   >>> f"{now}"
+   '2020-07-19 23:39:01.199675'
+
+Note that this is the :func:`str` format of a :mod:`datetime`. If you want to show the
+:func:`repr` format you use a conversion. There are three conversions beginning with the
+``'!'`` (exclamation) mark:
+
++------------+---------------+
+| Conversion | Meaning       |
++============+===============+
+| ``'!r'``   | :func:`repr`  |
++------------+---------------+
+| ``'!s'``   | :func:`str`   |
++------------+---------------+
+| ``'!a'``   | :func:`ascii` |
++------------+---------------+
+
+In our example of the datetime that becomes::
+
+   >>> f"{now!r}"
+   'datetime.datetime(2020, 7, 19, 23, 39, 1, 199675)'
+   >>> f"{now!a}"
+   'datetime.datetime(2020, 7, 19, 23, 39, 1, 199675)'
+
+While debugging it may be helpful to see both the expression and its value.
+This can be shown by appending the ``'='`` (equal) after the Python expression.
+
+::
+
+   >>> f"{now=}"
+   'now=datetime.datetime(2020, 7, 19, 23, 39, 1, 199675)'
+   >>> f"{ now = }"  # Note the addition of spaces
+   ' now = datetime.datetime(2020, 7, 19, 23, 39, 1, 199675)'
+
+Note that when using the debugging operator, any additional spaces will be left intact in the output.
+Also note that by default ``'='`` displays the :func:`repr` of the expression. It is possible to show
+the :func:`str` by appending a converter::
+
+   >>> f"{now=!s}"
+   'now=2020-07-19 23:39:01.199675'
+   >>> f"{ now = !a}" # Note the additional spaces
+   ' now = datetime.datetime(2020, 7, 19, 23, 39, 1, 199675)'
+
+Once the output has been evaluated it can then be formatted using the :func:`format` protocol.
+Formatting instructions are appended preceeded by a ``':'`` (colon).
+
+::
+
+   >>> f"{now:%B %d, %Y}"
```

Comment:
This format spec is a bit obscure, so I would use something simpler like alignment/padding on a string, or controlling the precision of a float like in the example below.

Reply:
Please look at how I have handled it now. It may enable easier understanding I think. Although it looks like the implementation of `__format__()` is not done well for all objects.